keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+    // IP addresses allocated to the pod. Routable at least within the cluster.
+    // Empty if not yet allocated.
+    // +optional
+    PodIPs []string `json:"podIP,omitempty" protobuf:"bytes,6,opt,name=podIP"`
+```
+This "PodIPs" slice representation should eventually become part of a versioned API (after a period of deprecation for the current versioned APIs).
+
+#### kubelet Startup Configuration for Multiple Pod CIDRs
+A new, plural "pod-cidrs" option for the [kubelet startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) is proposed, in addition to retaining the existing, singular "pod-cidr" option (for backwards compatibility):
+```
+  --pod-cidr   string        [Singular IP CIDR,  Default: ""]
+  --pod-cidrs  stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --pod-cidr argument will become deprecated.
+
+#### kube-proxy Startup Configuration for Multiple Pod CIDRs
+A new, plural "cluster-cidrs" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) is proposed, in addition to retaining the existing, singular "cluster-cidr" option (for backwards compatibility):
+```
+  --cluster-cidr   string        [Singular IP CIDR,  Default: ""]
+  --cluster-cidrs  stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --cluster-cidr argument will become deprecated.
+
+### Container Networking Interface (CNI) Plugin Considerations
+
+- This feature requires the use of CNI API 0.3.1 or later.
+- No changes to the CNI API specification (version 0.3.1 or later) are required for Kubernetes dual stack support.
+- A dual-stack capable CNI plugin should return pod IP assignments to kubelet via the "ips" array defined in the [CNI plugin result specification](https://github.com/containernetworking/cni/blob/master/SPEC.md#result).
+- It's expected that plugins will typically not assign more than one IPv4 address per pod.
+- The order of assigned IP addresses that are returned in the CNI plugin result for pod creation will be used to determine the "Default Pod IP" as described in the previous section.
+
+### Support for Multiple Service IPs per Service
+
+This proposal adds the capability to define multiple service CIDRs for a cluster.
+
+#### First Release of Dual Stack: No Service Address Configuration "Knobs"
+For the first release of Kubernetes dual-stack support, no new configuration "knobs" will be added for service definitions. This greatly simplifies the design and implementation, but requires imposing the following behavior:
+- Service IP allocation: Kubernetes will always allocate a service IP from each service CIDR for each service that is created. (In the future, we might want to consider adding configuration options to allow a user to select e.g. whether a given service should be assigned only IPv4, only IPv6, or both IPv4 and IPv6 service IPs.)
```

Comment:
If kube-proxy/plugins/etc are going to have to be updated to deal with multiple CIDR ranges anyway, then it would be useful to let the admin say "a.b.c.d/x should be treated as part of the service IP range, but the controller shouldn't allocate any new IPs out of that range". This would let you live-migrate the cluster from one service CIDR to another (something we (OpenShift) have had a handful of requests for).

(The same applies to the cluster CIDRs, though in that case the allocations are done by the plugin, not kube itself, so you could already implement this via plugin-specific configuration.)

Reply:
I have heard a VERY small number of similar requests, but I think it's orthogonal to multi-family.  We could add multi-CIDR with metadata (e.g. allow existing, but no more allocations) without touching multi-family support.

That said, *if* we do add multiple CIDRs, we should at least make it possible to add metadata later (so []struct, not []string).